### PR TITLE
Pre release 1.27.0

### PR DESCRIPTION
**Pull request by Joeper214**
## Features
- AWSpec test CRUD feature #338 
- CloudFormation Apply Date History #387 
- From TSD to Typings #366 
- Fetch Latest AWS regions #370 

## Bug fix
- Fix Detach EC2 Instance bug #389 

## Version Up
- Bundle Update on 2017-03-28 #385 
- Bundle Update on 2017-04-04 #388 
- Bundle Update on 2017-04-18 #390 

## Update process

``` sh
cd YOUR_SKYHOPPER_DIRECTORY
bower install
bundle exec rake db:migrate RAILS_ENV=production
bundle exec rake db:seed RAILS_ENV=production
cd frontend/
npm i
gulp ts
cd ..
```

### CODE REVIEW

File: app/models/node.rb [hunk 1]
```diff
@@ -187,36 +187,12 @@ def run_serverspec(infra_id, serverspec_ids, selected_auto_generated)
...
-          'success'
-        else
-          'pending'
-        end
-      else
-        'failed'
-      end
-
+    result =  generate_result(out)
+    Resource.find_by(physical_id: @name).status.servertest.update(value: result[:status_text])
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/controllers/servertests_controller.rb [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    @servertest = Servertest.new(infrastructure_id: infra_id)
+    @servertest.value = %!require '<choice_category>spec_helper'\n\n!
+  end
+
+  # GET /servertests/1
+  def show
+    render text: @servertest.value
+  end
+
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[brakeman]__
* Unescaped model attribute rendered inline [:ref](http://brakemanscanner.org/docs/warning_types/cross_site_scripting/)

File: app/models/servertest.rb [hunk 3]
```diff
@@ -8,14 +8,15 @@
...
-  has_many :dishes, through: :dish_serverspecs
-  has_many :resource_serverspecs
-  has_many :resources, through: :resource_serverspecs
-  has_many :serverspec_results, through: :serverspec_result_details
+  has_many :dish_servertests
+  has_many :dishes, through: :dish_servertests
+  has_many :resource_servertests
+  has_many :resources, through: :resource_servertests
+  has_many :servertest_results, through: :servertest_result_details
+  enum      category: [:awspec, :serverspec]
```
Comment (skyarch-networks) [inline app/models/servertest.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+class ServertestsController < ApplicationController
+  include Concerns::InfraLogger
+  class ServertestError < ::StandardError; end
+
+  before_action :set_servertest, only: [:update, :show, :edit, :destroy]
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  # GET /servertests/results
+  def results
+    physical_id = params.require(:physical_id)
+    infra_id    = params.require(:infra_id)
+    resource = Resource.where(infrastructure_id: infra_id).find_by(physical_id: physical_id)
+
+    @servertest_results = resource.servertest_results.order("created_at desc")
+
+    respond_to do |format|
+      format.json { render json: @servertest_results.as_json(only: [:id, :status, :message, :created_at, :category],
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        include: [{servertest_result_details: {only: [:id]}},{servertests: {only: [:name, :category]}}, {resource: {only: [:physical_id]}} ]) }
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: spec/jobs/servertest_job_spec.rb [hunk 7]
```diff
@@ -8,38 +8,38 @@
 
 require_relative '../spec_helper'
 
-RSpec.describe ServerspecJob, type: :job do
+RSpec.describe ServertestJob, type: :job do
```
Comment (skyarch-networks) [inline spec/jobs/servertest_job_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [76/25]

File: spec/models/servertest_spec.rb [hunk 8]
```diff
@@ -9,25 +9,25 @@
 require_relative '../spec_helper'
 
 # TODO: refactor
-describe Serverspec, type: :model do
-  let(:klass){Serverspec}
+describe Servertest, type: :model do
```
Comment (skyarch-networks) [inline spec/models/servertest_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [47/25]

File: spec/controllers/servertests_controller_spec.rb [hunk 9]
```diff
@@ -8,24 +8,25 @@
 
 require_relative '../spec_helper'
 
-describe ServerspecsController, type: :controller do
+describe ServertestsController, type: :controller do
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [340/25]

File: spec/controllers/servertests_controller_spec.rb [hunk 10]
```diff
@@ -8,24 +8,25 @@
...
     before do
-      create_list(:serverspec, 3, infrastructure_id: infrastructure.id)
+      create_list(:servertest, 3, infrastructure_id: infrastructure.id)
       req
     end
 
-    it 'should assign @serverspec' do
-      expect(assigns[:serverspecs]).to be_all{|serverspec|serverspec.kind_of?(Serverspec)}
+    it 'should assign @servertest' do
+      expect(assigns[:servertests]).to be_all{|servertest|servertest.kind_of?(Servertest)}
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Lint/AmbiguousBlockAssociation: Parenthesize the param `be_all{|servertest|servertest.kind_of?(Servertest)}` to make sure that the block will be associated with the `be_all` method call. [:ref](https://github.com/bbatsov/ruby-style-guide#syntax)

File: spec/controllers/servertests_controller_spec.rb [hunk 11]
```diff
@@ -45,8 +46,8 @@
         expect(assigns[:infrastructure_name]).to eq infrastructure.stack_name
       end
 
-      it 'should assign @serverspecs where infra id = infra.id(from url)' do
-        expect(assigns[:serverspecs]).to be_all{|serverspec|serverspec.infrastructure_id == infrastructure.id}
+      it 'should assign @servertests where infra id = infra.id(from url)' do
+        expect(assigns[:servertests]).to be_all{|servertest|servertest.infrastructure_id == infrastructure.id}
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Lint/AmbiguousBlockAssociation: Parenthesize the param `be_all{|servertest|servertest.infrastructure_id == infrastructure.id}` to make sure that the block will be associated with the `be_all` method call. [:ref](https://github.com/bbatsov/ruby-style-guide#syntax)

File: spec/controllers/servertests_controller_spec.rb [hunk 12]
```diff
@@ -282,29 +283,29 @@
...
 
-    it 'should assign @serverspec_results' do
-      expect(assigns[:serverspec_results]).to eq resource.serverspec_results
+    it 'should assign @servertest_results' do
+      expect(assigns[:servertest_results]).to eq resource.servertest_results
     end
   end
 
-  describe '#run' do
+  describe '#run_serverspec' do
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [63/25]

File: spec/policies/servertest_policy_spec.rb [hunk 13]
```diff
@@ -8,30 +8,30 @@
 
 require_relative '../spec_helper'
 
-describe ServerspecPolicy do
+describe ServertestPolicy do
```
Comment (skyarch-networks) [inline spec/policies/servertest_policy_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [51/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/386/diff/898ed862d225c322547f20a0041e3581d85c8e11?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

**Comment by Joeper214** [comment]
Can We Release Today? @root-beer

**Comment by root-beer** [comment]
Yes, please.

**Pull request merged**